serde_arrow/src/internal/serialization/outer_sequence_builder.rs

```diff
@@ -226,7 +227,52 @@ fn build_builder(path: String, field: &Field) -> Result<ArrayBuilder> {
                 .ctx(&ctx)?,
             )
         }
-        T::Struct(children) => A::Struct(build_struct(path, children, field.nullable)?),
+        T::Struct(children) => {
+            if let Some(Strategy::EnumsWithNamedFieldsAsStructs) =
+                get_strategy_from_metadata(&field.metadata)?
+            {
+                let mut related_fields: BTreeMap<&str, Vec<Field>> = BTreeMap::new();
+                let mut builders: Vec<(ArrayBuilder, FieldMeta)> = Vec::new();
+
+                for field in children {
+                    let Some(variant_name) = field.union_variant_name() else {
+                        // TODO: failure message
+                        continue;
+                    };
+
+                    let Some(field_name) = field.union_field_name() else {
+                        // TODO: failure message
+                        continue;
+                    };
+
+                    let mut new_field = field.clone();
+                    new_field.name = field_name;
+
+                    related_fields
+                        .entry(variant_name)
+                        .or_default()
+                        .push(new_field);
+                }
+
+                for (variant_name, fields) in related_fields {
```

Comment:
Why do you use two loops here? couldn't you simply build the builder where you push into `related_fields`? This way you also don't need to clone the fields.

Reply:
I have at aggregate/collect all the fields for that variant first, before creating the builder.